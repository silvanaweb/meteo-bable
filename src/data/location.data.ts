// This file provide the location DataCue, that means city and position.
// If the cities were a user input, we would have used an api call to a coordinate locator.
// Because we know the cities, we can hardcode their coordinates in a dictionary.
// We also use union types as helpers

export type Places = "london" | "paris" | "newyork" | "singapore" | "sydney";
export type Location = {
  name: string;
  latitude: number;
  longitude: number;
};
export type LocationsMap = { [key in Places]: Location };

const locations: LocationsMap = {
  london: {
    name: "London",
    latitude: 51.507351,
    longitude: -0.127758
  },
  paris: {
    name: "Paris",
    latitude: 48.856613,
    longitude: 2.352222
  },
  newyork: {
    name: "New York",
    latitude: 40.712776,
    longitude: -74.005974
  },
  singapore: {
    name: "Singapore",
    latitude: 1.28944,
    longitude: 103.849983
  },
  sydney: {
    name: "Sydney",
    latitude: -33.86882,
    longitude: 151.20929
  }
};

export function getLocations() {
  return { ...locations };
}
