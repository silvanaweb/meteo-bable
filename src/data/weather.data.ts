// This file provide the weather info. It fetches them from an API call.

export type WeatherInfo = {
  location: string;
  summary: string;
  temperature: number;
  icon: string;
};

export function getWeatherFromLocation(
  latitude: number,
  longitude: number
): Promise<WeatherInfo> {
  // simulate API call
  return new Promise(resolve => {
    setTimeout(() => {
      resolve({
        location: "London",
        summary: "Rainy days",
        temperature: 20,
        icon: "rain"
      });
    }, 2000);
  });
}
