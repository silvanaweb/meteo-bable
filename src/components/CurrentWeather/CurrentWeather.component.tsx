import React from "react";
import { PushSpinner } from "react-spinners-kit";

import { Selector, SelectorOption } from "../Selector/Selector.component";
import {
  WeatherView,
  WeatherViewType
} from "../WeatherView/WeatherView.component";
import {
  Location,
  getLocations,
  Places,
  getOptionsFromMap
} from "../../data/location.data";
import { getWeatherFromLocation } from "../../data/weather.data";
import { withRememberOption } from "../Helpers/RememberOption";
import { ErrorView } from "../ErrorView/ErrorView.component";
import { colors } from "../../theme/colors";

const SelectorWithMemory = withRememberOption(Selector);

type State = {
  locationOptions: Array<SelectorOption>;
  weatherInfo: WeatherViewType | null;
  loading: boolean;
  error: string | null;
};
type Props = {};

class CurrentWeather extends React.Component<Props, State> {
  private locationsMap = getLocations();

  constructor(props: Props) {
    super(props);
    this.state = {
      locationOptions: [],
      weatherInfo: null,
      loading: false,
      error: null
    };
  }

  componentDidMount() {
    const locationOptions = getOptionsFromMap(this.locationsMap);
    this.setState({ locationOptions });
  }

  render() {
    const { error, loading, locationOptions, weatherInfo } = this.state;
    return (
      <>
        <div className="pb--xl">
          <SelectorWithMemory
            loading={loading}
            onChange={this.handleLocationChange}
            options={locationOptions}
            title="Select a city"
          />
        </div>
        <ErrorView message={error}>
          <div style={WeatherContainerStyle}>
            <div style={spinnerStyle}>
              <PushSpinner size={50} color={colors.primary} loading={loading} />
            </div>
            <WeatherView weatherInfo={weatherInfo} loading={loading} />
          </div>
        </ErrorView>
      </>
    );
  }

  handleLocationChange = async (locationOption: SelectorOption) => {
    if (locationOption.value) {
      const locationKey: Places = locationOption.value as Places;
      const location: Location = this.locationsMap[locationKey];
      console.log(JSON.stringify(location));

      this.setState({ loading: true, error: null });
      try {
        // fetch and set weather info
        const info = await getWeatherFromLocation(
          location.latitude,
          location.longitude
        );
        this.setState({
          loading: false,
          weatherInfo: { ...info, location: location.name, id: locationKey }
        });
      } catch (error) {
        console.log(error);
        this.setState({
          error: "Weather forecast is not available now. Please, try later!",
          loading: false
        });
      }
    }
  };
}
const spinnerStyle: React.CSSProperties = {
  alignSelf: "center",
  left: "50%",
  position: "absolute",
  top: "50%",
  transform: "translate(-50%, -50%)",
  zIndex: 1
};
const WeatherContainerStyle: React.CSSProperties = {
  position: "relative",
  minHeight: 300,
  backgroundColor: "white",
  padding: "3rem 0"
};
export { CurrentWeather };
