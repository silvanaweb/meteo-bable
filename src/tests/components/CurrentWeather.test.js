import { shallow, ShallowWrapper } from "enzyme";
import toJSON from "enzyme-to-json";
import React from "react";
import { CurrentWeather } from "../../components/CurrentWeather/CurrentWeather.component";
import { optionsMock } from "../fixtures/location";
import * as WeatherData from "../../data/weather.data";

let wrapper;
jest.mock("../../data/weather.data");

describe("CurrentWeather", () => {
  beforeEach(() => {
    wrapper = shallow(<CurrentWeather />);
    wrapper.instance();
  });

  test("should render CurrentWeather correctly", () => {
    expect(toJSON(wrapper)).toMatchSnapshot();
  });

  test("should call handleLocationChange and update WeatherInfo state correctly", async () => {
    // mock the custom hook to have a desktop size
    const resp = {
      location: "london",
      temperature: 20,
      summary: "rainy days",
      icon: "rain"
    };
    const location = optionsMock[0];
    WeatherData.getWeatherFromLocation.mockResolvedValue(resp);
    await wrapper.instance().handleLocationChange(location);
    wrapper.update();
    expect(wrapper.state("weatherInfo")).toEqual({
      ...resp,
      location: location.label,
      id: location.value
    });
  });
});
