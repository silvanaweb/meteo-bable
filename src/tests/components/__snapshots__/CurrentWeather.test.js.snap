// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CurrentWeather should render CurrentWeather correctly 1`] = `
<Fragment>
  <div
    className="pb--xl"
  >
    <Component
      loading={false}
      onChange={[Function]}
      options={
        Array [
          Object {
            "label": "London",
            "value": "london",
          },
          Object {
            "label": "Paris",
            "value": "paris",
          },
          Object {
            "label": "New York",
            "value": "newyork",
          },
          Object {
            "label": "Singapore",
            "value": "singapore",
          },
          Object {
            "label": "Sydney",
            "value": "sydney",
          },
        ]
      }
      title="Select a city"
    />
  </div>
  <ErrorView
    message={null}
  >
    <div
      style={
        Object {
          "backgroundColor": "white",
          "minHeight": 300,
          "padding": "3rem 0",
          "position": "relative",
        }
      }
    >
      <div
        style={
          Object {
            "alignSelf": "center",
            "left": "50%",
            "position": "absolute",
            "top": "50%",
            "transform": "translate(-50%, -50%)",
            "zIndex": 1,
          }
        }
      >
        <Component
          color="#026565"
          loading={false}
          size={50}
          sizeUnit="px"
        />
      </div>
      <WeatherView
        loading={false}
        weatherInfo={null}
      />
    </div>
  </ErrorView>
</Fragment>
`;
